### ISSUE TITLE
Number element type doesn't validate properly for non-numeric or decimal input

### ISSUE THREAD

**Issue by bugfolder** [backdrop/backdrop-issues#5281]
**Description of the bug**

The `'number'` element type does not properly validate when given non-numeric or decimal numeric input.

**Steps To Reproduce**

To reproduce the behavior:

We'll make use of the fact that the EntityIDLayoutAccess class uses this element for the entity ID. The relevant code is:

```
    $form['entity_id'] = array(
      '#title' => t('@entity @entity_id_key is', array('@entity' => $this->entity_info['label'], '@entity_id_key' => strtoupper($this->entity_info['entity keys']['id']))),
      '#type' => 'number',
      '#min' => 1,
      '#default_value' => is_null($this->settings['entity_id']) ? NULL : (int) $this->settings['entity_id'],
      '#weight' => 100,
    );
```

(Incidentally, maybe that field should have `'#step' => 1` to ensure that it's an integer?)

Create a layout for path `node/%', then add a visibility condition of type "Node: NID", and try entering "foobar" or "5.678" as values. The first should be rejected as being non-numeric; the second should be accepted because it's a valid number (at least, in the usual sense of the word "number").

**Actual behavior**

Instead of rejecting "foobar" as being non-numeric, it is silently accepted and the value set to 0 (which is above the minimum value specified by `#min`).

Instead of accepting "5.678" as being a valid number, it is rejected with "Node NID is not a valid number."

**Additional information**

Add any other information that could help, such as:
- Backdrop CMS version: 1.20.x

**Comment by bugfolder** [comment]
The behavior is browser-dependent.

* Safari on Mac will let you type in "foobar", saves it as 0, but doesn't display any value.
* Same for Firefox on Mac.
* Chrome on Mac won't accept non-numeric keystrokes (it will accept . but not ,).

Part of this issue seems to be that the element is rendered by the [`<input type="number">` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number), whose implementation varies by browser (explaining the above), and which does some browser-level validation before passing its value on to the server. That explains why a string value like "foobar" isn't getting passed to the Backdrop form processor. The [browser-level validation](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number#validation) recognizes that that text is non-numeric, and so passes an empty string as the value of the input.

We don't seem to be making use of the browser-level validation.

**Comment by indigoxela** [comment]
@bugfolder if you dive into number form item validation, you might also be interested in #4751 (needs review).

I agree, the server-side validation for input type "number" needs some attention. Please note: only adding more to client-side validation seems insufficient to me.

**Comment by bugfolder** [comment]
> only adding more to client-side validation seems insufficient to me.

To me, too. In particular, client-side validation isn't always fully applied. It depends on whether the submission button is implemented as an `<input type="submit"...>` element or as something else, like `<button type="button"...>`.

Let's start with the former. If we create a standard form containing a number element, something like this:

```
function backdrop_test_form($form, &$form_state) {
  $form['backdrop_test_number'] = array(
    '#type' => 'number',
    '#title' => t('Number test'),
    '#description' => t('This is a number field with min=0, max=20, step=2.'),
    '#min' => 0,
    '#max' => 20,
    '#step' => 2,
  );
  $form['submit'] = array(
    '#type' => 'submit',
    '#value' => t('Submit'),
    '#suffix' => l(t('Cancel'), 'backdrop-test'),
  );
  return $form;
}
```

If we try to enter "foo" as the number value, we get varying behaviors, depending on the browser. See here (all on Mac):

![examples](https://user-images.githubusercontent.com/29214084/135763867-0793ce03-1c6d-44c4-95f9-28b54794a7c6.jpg)

* Safari will let us type in a number but when we try to submit it, it pops up the error "Enter a number". Note that this comes from Safari, not from Backdrop.
* Chrome simply won't accept a non-numeric keystroke, so the field remains blank.
* Firefox will let us type in a number, but when we try to submit it, it pops up the error "Please enter a number." (So Firefox is slightly more polite than Safari.)

Importantly, the submit button is rendered as this HTML:

```
<input type="submit" id="edit-submit" name="op" value="Submit" class="form-submit">
```

This matters because all submission buttons are not created equal. On the Layout visiblity dialog,

![layout_visibility](https://user-images.githubusercontent.com/29214084/135764024-d4dbeee8-ddb4-4abc-9a05-eee7852c9220.jpg)

the "Save visibility condition" button is rendered as this HTML:

```
<button type="button" class="button-primary form-submit ajax-processed ui-button ui-corner-all ui-widget">Save visibility condition</button>
```

Clicking this button doesn't seem to invoke the browser-level validation that blocks submission. That is, on Safari and Firefox, the submission is accepted. (Chrome, as before, doesn't accept the non-numeric keystrokes.)

However, both Safari and Firefox _do sanitize the submitted value_; in particular, instead of submitting "foo" they submit an empty string. (So what is saved is an empty value, but the next time the dialog is called up, the empty value gets converted to a 0 as the default value of the field.)

So, a couple of takeaways thus far:

* For forms that use an `<input type="submit"...>` element for submission, the browser will do its own validation blocking submission.
* For forms that don't, some browsers won't block submission, but will sanitize the submitted value.
* Non-numeric input on all browsers can be prevented at submission time by setting `'#required' => TRUE` in the form element, since that will reject the (empty) sanitized non-numeric value. However, the error message will be a bit cryptic to the user, because it will say "this field is required" rather than "the stuff you typed in was non-numeric."
* The [docs for the `<input>` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number) note that we still need to do server-side validation because a malicious user could circumvent the browser-level validation in what values actually get submitted.

**Comment by indigoxela** [comment]
@bugfolder the input element for visibility conditions in your screenshot is *no* number form element.

Text input fields don't get client side number validation. Can we please clarify, which form items should be number elements (input="number"), but currently still are textfields? Otherwise we cause confusion about what we're trying to fix.

**Comment by indigoxela** [comment]
Also, let's not try to "fix" client-side validation - that depends way too much on OS and browser. Example: FF on Linux marks non-numeric input as error and won't submit the form, Chromium refuses to add a non-numeric value (input stays empty). On Windows it's probably also different.

Values get submitted via POST, and IMO we should not blindly rely on client-side validation (depend on a "nice" client - not all clients are nice).

**Comment by bugfolder** [comment]
> the input element for visibility conditions in your screenshot is no number form element.

Sorry, screenshot was out of date. Here's the current version. (A PR was recently merged that converted the entity ID to a `number` type.) Everything I said was applicable to the `number` type.

![visibility](https://user-images.githubusercontent.com/29214084/135884580-8ba5b6e9-ec73-4c26-ad88-3a1a9fd881f5.jpg)

**Comment by bugfolder** [comment]
> Also, let's not try to "fix" client-side validation - that depends way too much on OS and browser. 

Yeah, that was my point. We can't "fix" the fact that Chrome prevents submission of a non-numeric Node NID on this form, while Safari allows submission but converts it to an empty string. We can, however, on the server side handle those differences to some degree, e.g., PR https://github.com/backdrop/backdrop/pull/3771.
